src/els_diagnostics_server.erl

```diff
@@ -0,0 +1,168 @@
+%%==============================================================================
+-type state() :: #state{}.
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-spec start_link({atom(), uri()}, term()) -> {ok, pid()}.
+start_link({Id, Uri}, _Opts) ->
+  {ok, _} = gen_server:start_link({local, Id},
+                                  ?MODULE,
+                                  {Id, Uri},
+                                  []).
+
+%%==============================================================================
+%% gen_server callbacks
+%%==============================================================================
+-spec init({atom(), uri()}) -> {ok, state()}.
+init({Id, Uri}) ->
+  State = #state{ server_name = Id
+                , uri = Uri
+                },
+  {ok, State}.
+
+-spec handle_call(any(), any(), state()) -> {reply, any(), state()}.
+handle_call(get_diagnostics, _From, #state{diagnostics = Diagnostics} = S) ->
+  {reply, Diagnostics, S};
+handle_call(UnknownMessage, _From, S) ->
+  lager:info("Unknown handle_call! message: ~p", [UnknownMessage]),
+  {reply, ok, S}.
+
+-spec handle_cast(any(), state()) -> {noreply, state()}.
+handle_cast(run, #state{uri = Uri} = S) ->
+  Diagnostics =
+    lists:foldl(
+      fun(Mod, AccumulatedDiagnostics0) ->
+          Diagnostics = diagnostic(Mod, Uri),
+          AccumulatedDiagnostics = AccumulatedDiagnostics0 ++ Diagnostics,
+          send_notification(Uri, AccumulatedDiagnostics),
+          AccumulatedDiagnostics
+      end, [], [ els_compiler_diagnostics
+               , els_elvis_diagnostics
+               , els_dialyzer_diagnostics
+               ]),
+  {noreply, S#state{diagnostics = Diagnostics}};
+handle_cast(UnknownMessage, S) ->
+  lager:info("Unknown handle_cast! message: ~p", [UnknownMessage]),
+  {noreply, S}.
+
+-spec handle_info(any(), state()) -> {noreply, state()}.
+handle_info({'EXIT', _, normal}, S) ->
+  {noreply, S};
+handle_info(UnknownMessage, S) ->
+  lager:info("Unknown handle_info! message: ~p", [UnknownMessage]),
+  {noreply, S}.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+-spec on_save(uri()) -> ok.
+on_save(Uri) ->
+  els_diagnostics_sup:on_save(Uri),
+  Id = binary_to_atom(Uri, utf8),
+  ok = gen_server:cast(Id, run).
+
+-spec on_open(uri()) -> ok.
+on_open(Uri) ->
+  els_diagnostics_sup:on_open(Uri),
+  Id = binary_to_atom(Uri, utf8),
+  case gen_server:call(Id, get_diagnostics, 100) of
+    undefined -> gen_server:cast(Id, run);
+    Diagnostics -> Diagnostics
+  end.
+
+-spec diagnostic(atom(), uri()) -> [diagnostic()].
+diagnostic(els_compiler_diagnostics, Uri) ->
+  CDiagnostics = els_compiler_diagnostics:diagnostics(Uri),
+  maybe_compile_and_load(Uri, CDiagnostics),
```

Comment:
Eventually I would really like to remove this special case for the `els_compiler_diagnostics`. Ideally there would be no coupling between these two, otherwise we could just move the whole `maybe_compile_and_load/2` function into the `els_compiler_diagnostics`.

Reply:
I agree, and seeing as I'd like to extend the "do-something-on-a-remote-node"functionality that I introduced previously, I think we should have a re-thinking of this. Are you fine with discussing this in an issue separate of this PR?